### DISCUSSION TITLE
Stable AIR SDK for annual Android updates? (and updated ANE requirements)

### DISCUSSION THREAD

**Discussion by FliplineStudios** [airsdk/Adobe-Runtime-Support#2673] [Q&A]
With the Google Play deadline approaching for updating the target API for all existing apps, we're getting ready to update a couple dozen Android apps that were previously targeting API 30 and using legacy builds.  I was wondering if anyone has suggestions of the best stable version of the AIR SDK that we should be using nowadays for the updates.  All of our live apps were last built with 33.1.1.743 last year, so I'm assuming we should be switching to something newer than that.  With seeing recent issue reports of crashes with 50.2.2.x I wasn't sure whether it would be safe to use these newer versions for production, or if we should stick to the 33.1 branch, or if there's an older 50.x that should be used instead (or maybe there's another 50.x update on the horizon that we should wait for).  Getting through all of the game updates (with updating ANEs, and playtesting on multiple devices) will be quite a few weeks of work, so I'd prefer to just go through the process once with something stable and not have to keep updating :)

I'd love to just update to API Level 33 so we can at least skip a year with these annual Google Play updates, but if we should stick to an older SDK and use API 31 if that's more stable that's also an option.  All of our apps are GPU-rendermode, don't use Workers, and don't use the newer running in background thread, if that makes a difference on what's safe to be using in that case.  Also wasn't sure if we could just target API 33 in the older 33.1 SDKs, or if it's only safe to be targeting API level 33 with the newer 50.x SDKs...

We only use Distriqt ANEs (mainly InAppBilling and Application, and one app uses Share and CameraRollExtended), so we'll also have to update those.  This may be mainly a question for @marchbold, but it sounds like there have been some changes in the requirements for using Distriqt ANEs as well since last year.  I didn't notice the information in the docs for the ANEs, but from reading through the AIR issue tracker it sounds like these requirements have changed?

- Can't use buildLegacyAPK anymore? (currently using this for 26+ apps)
- Need to use at least swf-version 44 in app's main SWF? (currently using swf-version 35 or 37 in all apps)
- Is there a specific Android SDK required that we'll need to set up in Android Studio?

If anyone has any advice based on what you've gone live with recently that seems stable, I'm open to any suggestions!  I'll also keep checking the SDK release notes and AIR issue tracker to see what else we might need to look out for.

**Comment by ajwfrost** [comment]
> maybe there's another 50.x update on the horizon that we should wait for

Yes, hopefully the next one will resolve some of these stability issues on the Android side, we've got a number of fixes going in.
Not using Workers and background threads is definitely helpful..

I don't see anything wrong with setting your target API level to 33 in an older AIR SDK build though. So that would probably be the lowest risk approach.

The newer ANEs do also need the new build mechanism using Gradle; I'm not sure if you can get away with just uploading individual APK files still? If you can, then fine, best to stick with the set-up you have and just change the target SDK level. But otherwise, you'd need to update everything to use Gradle builds and can then use the newer ANEs.

I don't think there's a need to use a SWF version of 44 (unless perhaps this is a requirement from the ANE side?) - and for the Android SDK, that can be the latest one, it just has to support API 33..

Might be worth seeing if you can get away with the existing set-up with the updated target SDK level, and then in parallel look at taking the next release and updating all the ANEs etc..?

**Comment by FliplineStudios** [comment]
Thanks for the feedback Andrew!  At least for legacy apps they still allow separate APKs, as long as they were published before AABs became a requirement they've kept allowing APKs for updated builds (at least they did as of late last year).  For in-app purchases we do need to use a newer Play Billing Library version though due to those changing annual requirements too (using v3 right now), so pretty sure we'll at least need to use the newer ANEs and switch to the Gradle builds.  Unless the ANE that started with v5 still allowed legacy builds and I can grab that older version, I'm not sure when exactly the ANE requirements changed.

Some of these apps started back with AIR 3.1, so we try not to change things too much with the build processes if we can help it -- which was a lot easier before Google started requiring annual updates :)  

We'll at least get our Android SDK updated in Android Studio to API 33, and see how well it works targeting that with 33.1 -- I'm assuming we should at least use the latest 33.1.1.935 to handle the OpenSSL update?

**Reply by ajwfrost** [reply]
Ah, I forgot about the OpenSSL requirement, so yes, 33.1.1.935 would be the minimum I guess..

**Comment by marchbold** [comment]
Regarding the ANEs, we definitely no longer support the legacy build method and haven't for a while now. I wouldn't suggest trying to release a new app using this method as you'll be using very old versions of ANEs and wouldn't likely pass review.

Additionally we have started moving to the even newer gradle dependencies for some of our extensions, this requires 50.2. This is due to some of the newer Android dependencies not working with the older approach. Currently there is only one dependency using this `com.google.code.gson` so if the extensions you are using don't require this one, then you should be okay to continue on with 33.1.1.935. 

If you are only using:
- `com.distriqt.InAppBilling` 
- `com.distriqt.Share`
- `com.distriqt.Application`
- `com.distriqt.CameraRollExtended`

then you should be fine to continue using 33.1.1.935.

**Reply by FliplineStudios** [reply]
Thanks!  We last used InAppBilling v13.1.0 which was already pretty old at the time (though worked fine with legacy apks and swf-version 37), I figured with the newer ANEs we'd likely have to switch over to the regular gradle builds for all of these updates.

Looking at the extension XMLs, it looks like we do need to use at least a swf-version of 44 though?

**Reply by marchbold** [reply]
Yeah you should use swf version 44, but that is what is in 33.1.1.935. Any reason you don't want to update the swf version?

**Reply by FliplineStudios** [reply]
We've just always built the SWF separately with one set of tools, and then later compiled it from the command line with whatever the latest AIR version is -- sometimes keeping the same SWF and just grabbing the newest AIR SDK for the final compile to apk/ipa/etc.  

Also we do all of the development of our games on Windows, so especially once swf-version 38+ started affecting GPU-rendering performance from the Direct3D changes for Windows, we've just kept it at 37 to make development go smoother.  It didn't really have any effect on the final apps (and we didn't need access to features from later AIR versions), so we've just kept that same pipeline of development/testing/building the SWF with an older version, and compiling to app with a newer AIR SDK later.

Guess we'll need to change up our pipeline a bit now :)  Thankfully the maxD3D tag should help with avoiding performance issues during early desktop testing at least.

**Reply by idanasher** [reply]
> com.google.code.gson

and if I do use com.google.code.gson
and I want the latest ane's ( i use a LOT of them) - so which stable AIR version do I need to use ?

**Reply by ajwfrost** [reply]
If you're using that, which uses the 'gradle' based dependency, then it needs to be 50.2.1.1 or higher. This, or 50.2.2.2, might be the best bets. If you're seeing any issues with using the storage access framework on Android with that, you can disable it using the `storageAccessFrameworkFromAPI` app descriptor setting set to 999.

But if you're able to hold on for a week or two, we expect 50.2.3.1 to then come out and this should improve the stability...

**Reply by idanasher** [reply]
Thanks Andrew
meanwhile, I have managed to upload a new versopn of my my app with 33.1.1.935 and some latest's distriqt ane's (facebook, core, inapp & adverts) and the all the rest (including the 'com.google.code.gson') - stayed as before, with much older versions.

so far it looks good but I hope not to see any crashes movement on the graph
Thanks Andrew & @marchbold

**Reply by FliplineStudios** [reply]
@idanasher Good to know that that works!  Did you target Android API Level 33 too?  Or targeting 31?

**Comment by idanasher** [comment]
targeting 31, still
when is due date for targeting 33 ?

**Reply by FliplineStudios** [reply]
After August 31st, both new apps and updates to existing apps need to target 33.  Existing apps in the Play Store just need to be targeting 31 by then though, and then they'll be OK for another year as long as you don't need to update the app after that (and if you did need to update over the next year, you'd have to switch to target 33 then).  And then next year all of the existing apps will need to target 33 to be able to be installed on current devices, and the cycle continues...
